File: Graph/ts/dfs.ts
import { Graph } from "./types";

export function depthFirstTraversal(graph: Graph, source: string): string[] {
  if (!source) return [];

  const order: string[] = [];
  const stack: string[] = [source];
  const visited = new Set<string>();

  while (stack.length > 0) {
    const current = stack.pop()!;
    if (visited.has(current)) continue;
    visited.add(current);
    order.push(current);

    const neighbors = graph[current] ?? [];
    for (let i = neighbors.length - 1; i >= 0; i -= 1) {
      stack.push(neighbors[i]);
    }
  }

  return order;
}

File: Graph/go/dfs.go
package graph

// DepthFirstTraversal visits nodes in depth-first order using an explicit stack.
func DepthFirstTraversal(g Graph, source string) []string {
	if source == "" {
		return []string{}
	}

	order := []string{}
	stack := []string{source}
	visited := map[string]struct{}{}

	for len(stack) > 0 {
		idx := len(stack) - 1
		node := stack[idx]
		stack = stack[:idx]

		if _, seen := visited[node]; seen {
			continue
		}
		visited[node] = struct{}{}
		order = append(order, node)

		neighbors := g[node]
		for i := len(neighbors) - 1; i >= 0; i-- {
			stack = append(stack, neighbors[i])
		}
	}

	return order
}
